api/src/model/request/play_asset_pack_fetch_request.gd

```diff
@@ -0,0 +1,127 @@
+
+var _pack_name : String
+var _state : PlayAssetPackState
+var _error : PlayAssetPackException
+
+# used for unit testing
+var _stub_play_asset_pack_manager
+
+func _init(pack_name):
+	_pack_name = pack_name
+	
+	# _state will be defaulted with status of UNKNOWN
+	var default_pack_dict = {
+		PlayAssetPackState._NAME_KEY: pack_name, 
+		PlayAssetPackState._STATUS_KEY: PlayAssetPackManager.AssetPackStatus.UNKNOWN, 
+		PlayAssetPackState._ERROR_CODE_KEY: PlayAssetPackManager.AssetPackErrorCode.NO_ERROR,
+		PlayAssetPackState._BYTES_DOWNLOADED_KEY: 0,
+		PlayAssetPackState._TOTAL_BYTES_TO_DOWNLOAD_KEY: 0,
+		PlayAssetPackState._TRANSFER_PROGRESS_PERCENTAGE_KEY: 0
+	}
+	_state = PlayAssetPackState.new(default_pack_dict)
+	
+# -----------------------------------------------------------------------------
+# Returns the requested asset pack's name.
+# -----------------------------------------------------------------------------
+func get_pack_name() -> String:
+	return _pack_name
+
+# -----------------------------------------------------------------------------
+# Returns whether this request is completed or not.
+# -----------------------------------------------------------------------------
+func get_is_completed() -> bool:
+	var is_non_existing_pack_exception = _state.get_status() == PlayAssetPackManager.AssetPackStatus.UNKNOWN \
+		and _state.get_error_code() == PlayAssetPackManager.AssetPackErrorCode.INTERNAL_ERROR
+	var is_terminal_state = _state.get_status() in PlayAssetPackManager._PACK_TERMINAL_STATES
+	
+	return is_terminal_state or is_non_existing_pack_exception
+
+# -----------------------------------------------------------------------------
+# Returns the most up-to-date PlayAssetPackState object.
+# -----------------------------------------------------------------------------
+func get_state() -> PlayAssetPackState:
+	return _state
+
+# -----------------------------------------------------------------------------
+# Returns a PlayAssetPackException if exception occurred.
+# -----------------------------------------------------------------------------
+func get_error() -> PlayAssetPackException:
+	return _error
+
+# -----------------------------------------------------------------------------
+# Callback functions handling signals emitted from the plugin.
+# -----------------------------------------------------------------------------
+func _on_fetch_success(result: Dictionary):
+	# Since fetch() in plugin returns a PlayAssetPackStates Dictionary, we need to extract
+	# the PlayAssetPackState within.
+	var fetch_asset_pack_states_dict = PlayAssetPackStates.new(result).get_pack_states()
+	
+	if fetch_asset_pack_states_dict.has(_pack_name):
+		_on_state_updated(fetch_asset_pack_states_dict[_pack_name].to_dict())
+		# emit non-duplicating state_updated signal
+		if _stub_play_asset_pack_manager == null:
+			PlayAssetPackManager._forward_high_level_state_updated_signal(_pack_name, _state.to_dict())
```

Comment:
This logic might make more sense inside of _forward_fetch_success and _forward_fetch_error so that it stays in the PlayAssetPackManager. That way logic/state flows in only one direction: from PlayAssetDeliveryManager to the request.

Reply:
Yeah I think untangle this logic actually make the readability of this much better. 
In order to get the assetPackState for fetchError I used ```	var previous_state = target_request.get_state().to_dict() ``` to get the previous state.